[FLINK-32492][table-planner] Introduce FlinkRuntimeFilterProgram to apply the runtime filter optimization.

wanglijie95 opened the pull request:
## What is the purpose of the change
Introduce FlinkRuntimeFilterProgram to apply the runtime filter optimization.

## Verifying this change
Add test `FlinkRuntimeFilterProgramTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**no**)
  - The serializers: (**no**)
  - The runtime per-record code paths (performance sensitive): (**no**)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**no**)
  - The S3 file system connector: (**no**)

## Documentation

  - Does this pull request introduce a new feature? (**yes**)
  - If yes, how is the feature documented? (**not documented**)

wanglijie95 commented:
cc @lsyldliu

wanglijie95 commented:
Could you help to review this PR when you have time @swuferhong

lsyldliu commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/runtimefilter/BatchExecLocalRuntimeFilterBuilder.java:
```diff
@@ -0,0 +1,85 @@
+    @SuppressWarnings("unchecked")
+    protected Transformation<RowData> translateToPlanInternal(
+            PlannerBase planner, ExecNodeConfig config) {
+        ExecEdge inputEdge = getInputEdges().get(0);
```
Throw exception directly.

wanglijie95 replied:
Fixed

lsyldliu commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/runtimefilter/BatchExecRuntimeFilter.java:
```diff
@@ -0,0 +1,83 @@
+    @SuppressWarnings("unchecked")
+    protected Transformation<RowData> translateToPlanInternal(
+            PlannerBase planner, ExecNodeConfig config) {
+        ExecEdge leftInputEdge = getInputEdges().get(0);
```
Ditto

wanglijie95 replied:
Fixed

lsyldliu commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/physical/batch/runtimefilter/BatchPhysicalGlobalRuntimeFilterBuilder.java:
```diff
@@ -0,0 +1,78 @@
+ * Batch physical RelNode responsible for aggregating all received filters into a global filter. See
+ * {@link FlinkRuntimeFilterProgram} for more info.
+ */
+public class BatchPhysicalGlobalRuntimeFilterBuilder extends SingleRel implements BatchPhysicalRel {
```
For the runtime filter builder operator, can we reuse the hashagg operator?

wanglijie95 replied:
For simplicity, I prefer to do it in the future.

swuferhong commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgram.java:
```diff
@@ -0,0 +1,571 @@
+                    buildSideChecker);
+
+        } else {
+            // more cases
```
We need to verify if we have at least covered all the TPC-DS cases that can apply runtime filters.

wanglijie95 replied:
TPCDS-related tests are still in progress

swuferhong commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgram.java:
```diff
@@ -0,0 +1,571 @@
+                    program.getProjectList().stream()
+                            .map(program::expandLocalRef)
+                            .collect(Collectors.toList());
+            ImmutableIntList inputIndices = getInputIndices(projects, buildIndices);
```
Can this method be reused with DPP utils?

wanglijie95 replied:
No, the logic is not exactly same. In runtime filter, the return value of `getInputIndices` is either an empty list(which means the push-down is failed) or a list of the same length as `buildIndices`(because it requires a one-to-one correspondence between the build-indices and probe-indices). But in DPP, it does not have this requirement.

swuferhong commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgram.java:
```diff
@@ -0,0 +1,571 @@
+
+        } else if (rel instanceof Join) {
+            Join join = (Join) rel;
+            if (!(join.getLeft() instanceof Exchange) && !(join.getRight() instanceof Exchange)) {
+                // TODO: Is there such case?
+                return Optional.empty();
```
The coverage pattern here should be added on demand, so we can leave it for now. Later, we can provide PR support to add coverage after TPC-DS verification.

swuferhong commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgramTest.java:
```diff
@@ -0,0 +1,271 @@
+    private final TestValuesCatalog catalog =
+            new TestValuesCatalog("testCatalog", "test_database", true);
+
+    @Before
```
Using Junit5 for junit4-free item?

wanglijie95 replied:
We have to use junit4 here, because the base class `TableTestBase` is based on junit4.

swuferhong commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgramTest.java:
```diff
@@ -0,0 +1,271 @@
+                OptimizerConfigOptions.TABLE_OPTIMIZER_RUNTIME_FILTER_MIN_PROBE_DATA_SIZE,
+                MemorySize.parse("10g"));
+        tableConfig.set(
+                OptimizerConfigOptions.TABLE_OPTIMIZER_RUNTIME_FILTER_MIN_FILTER_RATIO, 0.5);
```
why need this config?

wanglijie95 replied:
I explicitly set the runtime-filter related config options here, due to the default value may be changed later(The best default values are still in testing). We explicitly set them here, even if the default values are changed later, current test will not be affected.

swuferhong commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgramTest.java:
```diff
@@ -0,0 +1,271 @@
+                OptimizerConfigOptions.TABLE_OPTIMIZER_RUNTIME_FILTER_MIN_FILTER_RATIO, 0.5);
+        util.getTableEnv()
+                .getConfig()
+                .set(OptimizerConfigOptions.TABLE_OPTIMIZER_BROADCAST_JOIN_THRESHOLD, -1L);
```
I think this config need not to add in before() method.

wanglijie95 replied:
Yes, you are right. The `TABLE_OPTIMIZER_BROADCAST_JOIN_THRESHOLD` and `TABLE_OPTIMIZER_AGG_PHASE_STRATEGY` are used to let the build side is a direct Agg or Calc (without Exchange). I move these 2 options to the test `testBuildSideIsAggWithoutExchange` and `testBuildSideIsCalcWithoutExchange`

swuferhong commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgramTest.java:
```diff
@@ -0,0 +1,271 @@
+        util.verifyRelPlan(query);
+        util.verifyExecPlan(query);
+    }
+
```
Can you provide an example of a multi-level join?

wanglijie95 replied:
The test `testBuildSideIsJoinWithoutExchange` is the case of multi-level join.

swuferhong commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgramTest.java:
```diff
@@ -0,0 +1,271 @@
+        util.verifyExecPlan(query);
+    }
+
+    @Test
```
I have one question: do we need to consider conflicts between runtime filters and DPP for partitioned tables?

wanglijie95 replied:
I think the runtime filter doest not conflict with DPP, they can exist at the same time. WDYT ?

swuferhong replied:
> I think the runtime filter doest not conflict with DPP, they can exist at the same time. WDYT ?

Yes, I understand now

swuferhong commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgramTest.java:
```diff
@@ -0,0 +1,271 @@
+        String query =
+                "select * from dim, fact, fact2 where fact.amount = fact2.amount and"
+                        + " fact.amount = dim.amount and dim.price < 500";
+        util.verifyRelPlan(query);
```
Only need to verify ExecPlan, no need to verify twice.

wanglijie95 replied:
I hope to verify the `ast`, `opt rel` and `opt exec`. I change it to call the `verifyPlan` only.

swuferhong reviewed (commented):
Hi @wanglijie95 . Thanks for your contribution, I left some comments.

swuferhong commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/OptimizerConfigOptions.java:
```diff
@@ -165,6 +167,51 @@ public class OptimizerConfigOptions {
+
+    @Documentation.TableOption(execMode = Documentation.ExecMode.BATCH)
+    public static final ConfigOption<MemorySize>
+            TABLE_OPTIMIZER_RUNTIME_FILTER_MAX_BUILD_DATA_SIZE =
+                    key("table.optimizer.runtime-filter.max-build-data-size")
+                            .memoryType()
+                            .defaultValue(MemorySize.parse("10m"))
```
> What do you rely on to determine these default values? Is there any relevant trade off performance analysis? BTW, can you explain why we need the minimum and maximum values here?

explain in comments.

wanglijie95 replied:
I 've added comments for the 3 related config options, PTAL.

wanglijie95 replied:
The best vaule of default are still in testing. I will try to find a best default value based on the performance result of TPCDS-10T.

wanglijie95 commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/OptimizerConfigOptions.java:
```diff
@@ -165,6 +167,51 @@ public class OptimizerConfigOptions {
+    public static final ConfigOption<Boolean> TABLE_OPTIMIZER_RUNTIME_FILTER_ENABLED =
+            key("table.optimizer.runtime-filter.enabled")
+                    .booleanType()
+                    .defaultValue(false)
```
We hope to enable this feature by default in future releases. But in 1.18, it will be disabled by default.

wanglijie95 commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/OptimizerConfigOptions.java:
```diff
@@ -165,6 +167,51 @@ public class OptimizerConfigOptions {
+                            .memoryType()
+                            .defaultValue(MemorySize.parse("10m"))
+                            .withDescription(
+                                    "Data volume threshold of the runtime filter build side. "
```
Fixed

wanglijie95 commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/config/OptimizerConfigOptions.java:
```diff
@@ -165,6 +167,51 @@ public class OptimizerConfigOptions {
+                            .withDescription(
+                                    Description.builder()
+                                            .text(
+                                                    "Data volume threshold of the runtime filter probe side. "
```
Fixed

wanglijie95 commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgram.java:
```diff
@@ -0,0 +1,571 @@
+        }
+    }
+
+    private static Optional<Double> getEstimatedDataSize(RelNode relNode) {
```
Done

wanglijie95 commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgram.java:
```diff
@@ -0,0 +1,571 @@
+
+    @Override
+    public RelNode optimize(RelNode root, BatchOptimizeContext context) {
+        final RuntimeFilterConfig config =
```
I agree with you. Done.

wanglijie95 commented:
Thanks for review @lsyldliu @swuferhong . All comments have been addressed or replied ,PTAL.

swuferhong commented:
> Thanks for review @lsyldliu @swuferhong . All comments have been addressed or replied ,PTAL.

Hi, Thanks! @wanglijie95 . Now LGTM for this PR.  I wish the itcase can be supplemented after all PR development is completed, And after the TPC-DS testing is completed, more test patterns can be added.

lsyldliu commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgram.java:
```diff
@@ -0,0 +1,563 @@
+     * @return An optional info of the suitable build side.It will be empty if we cannot find the
+     *     suitable build side.
+     */
+    private static Optional<BuildSideInfo> findSuitableBuildSide(
```
Can you give more explanation of this method's logic? Especially the various if else branch.

lsyldliu commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgram.java:
```diff
@@ -0,0 +1,563 @@
+        return ImmutableIntList.copyOf(inputIndices);
+    }
+
+    private static Tuple2<ImmutableIntList, ImmutableIntList> getInputIndices(
```
Give a java doc explain this method logic?

lsyldliu commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgram.java:
```diff
@@ -0,0 +1,563 @@
+                            ignoreExchange(probeSide),
+                            suitableBuildInfo.buildIndices,
+                            probeIndices);
+            if (probeSide instanceof Exchange) {
```
IMO, the current logic is that we inject the runtime filter node regardless of whether the probe side is an exchange node or not, can you give some comments to explain this?

lsyldliu reviewed (commented):
@wanglijie95 Thanks for your contribution, I left some comments.

wanglijie95 commented:
Thanks for comments @lsyldliu. I 've addressed all comments, PTAL. Besides, I add probe side push down logic in the last commit, to get better performance

lsyldliu commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgram.java:
```diff
@@ -0,0 +1,678 @@
+            // try to push the probe side to input of agg, iff the indices are all in grouping keys.
+            BatchPhysicalGroupAggregateBase agg = (BatchPhysicalGroupAggregateBase) rel;
+            int[] grouping = agg.grouping();
+            if (probeIndices.stream().allMatch(index -> (index < grouping.length))) {
```
It seems that this judge's logic is not correct? The `index` is not necessarily smaller than `grouping.length`.

wanglijie95 replied:
The first `n(grouping.length)` elements of output of group agg are the group keys. In other words, if the index is greater than the `grouping.length`, it means it's not in the group key set, and cannot be pushed down further.

DPP  has the  [`similar logic`](https://github.com/apache/flink/blob/17be3716be42cd9dc5986e97ae222971d9d19be7/flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java#L414C1-L415C1)

Correct me if I'm wrong.

lsyldliu replied:
Yes, you are right

lsyldliu commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgramTest.java:
```diff
@@ -0,0 +1,417 @@
+import java.util.Collections;
+
+/** Test for {@link FlinkRuntimeFilterProgram}. */
+public class FlinkRuntimeFilterProgramTest extends TableTestBase {
```
Please also add some ITCase to verify the overall process.

wanglijie95 replied:
Added

lsyldliu reviewed (commented):
Thanks for contribution, Looks good overall, just left two comments.

wanglijie95 commented:
@lsyldliu I 've addressed or replied your comments, PTAL.

lsyldliu commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/RuntimeFilterITCase.java:
```diff
@@ -0,0 +1,213 @@
+                .set(
+                        OptimizerConfigOptions.TABLE_OPTIMIZER_RUNTIME_FILTER_MAX_BUILD_DATA_SIZE,
+                        MemorySize.parse("10m"));
+        tEnv.getConfig()
```
Don't need to set `TABLE_OPTIMIZER_RUNTIME_FILTER_MIN_PROBE_DATA_SIZE` and `TABLE_OPTIMIZER_RUNTIME_FILTER_MIN_FILTER_RATIO`, it is equal to the default value.

wanglijie95 replied:
Fixed

lsyldliu commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/RuntimeFilterITCase.java:
```diff
@@ -0,0 +1,213 @@
+
+    @ParameterizedTest(name = "mode = {0}")
+    @MethodSource("parameters")
+    void testSimpleRuntimeFilter(BatchShuffleMode shuffleMode) {
```
Do you have check the plan whether is as expected?

wanglijie95 replied:
Yes, have checked.

lsyldliu commented on flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgramTest.xml:
```diff
@@ -0,0 +1,1022 @@
+		</Resource>
+	</TestCase>
+
+
```
Why are there some blank lines here?

wanglijie95 replied:
Delelted

lsyldliu commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgram.java:
```diff
@@ -0,0 +1,678 @@
+                return join.copy(join.getTraitSet(), Arrays.asList(leftSide, rightSide));
+            }
+        } else if (rel instanceof BatchPhysicalGroupAggregateBase) {
+            // try to push the probe side to input of agg, iff the indices are all in grouping keys.
```
typo -> if

wanglijie95 replied:
Fixed

lsyldliu reviewed (commented):
Thanks for update, I left some comments

lsyldliu reviewed (approved):
LGTM

wanglijie95 commented:
Thanks for review @lsyldliu @swuferhong ! I'll merge this PR once the CI done.
